### ISSUE TITLE
Incorrect usage of Service Bus clients can lead to socket exhaustion

### ISSUE THREAD

**Issue by mattbrooks2010** [EnableSoftware/Enable.Extensions.Queuing#20]
# Problem

Incorrect usage of Service Bus clients can lead to socket exhaustion.

The library doesn't treat Service Bus client types such as `MessageSender` and `MessageReceiver` as singletons. Every instance of `AzureServiceBusQueueClient` created by an application will also create instances of `MessageSender` and `MessageReceiver`. 

An application could of course mitigate this directly by using `AzureServiceBusQueueClient` as a singleton. However, many applications are already creating an instance of this type on a 'per use' basis.

```c#
using (var queue = queueClientFactory.GetQueueReference("test-queue"))
{
    await queue.EnqueueAsync(message);
}
```

# Suggestion

Because it is already very common for applications to use the pattern above, it would be beneficial if this library could mitigate this problem directly. It could do this by internally caching instances of `AzureServiceBusQueueClient`, or the `MessageSender` and `MessageReceiver` it creates.

# References

- https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements#reusing-factories-and-clients

- https://github.com/Azure/azure-service-bus-dotnet/issues/130#issuecomment-294946723

**Comment by mattbrooks2010** [comment]
Following some internal dialogue, the agreed tasks are:

- [ ] Implement a cache of `AzureServiceBusQueueClient` instances. We'd like each `AzureServiceBusQueueClientFactory` to maintain its own object cache, keyed by `queueName`. 
- [ ] Remove current implementation of `Dispose` in `AzureServiceBusQueueClient`. 
- [ ] Add unit tests that verify the `AzureServiceBusQueueClient` instances returned by `GetQueueReference` are correctly cached by `queueName`.
- [ ] Integrate a test version of this package with an existing solution and test.
- [ ] Publish a new release version of this library to NuGet.

**Comment by Peter554** [comment]
Given that we have a similar issue for RabbitMQ (#21) and that the [Microsoft docs](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements#reusing-factories-and-clients) discourage disposing of QueueClients perhaps we ought to also more gradually change the default abstraction pattern to not use a `using` clause (as encouraged in the README)? We would need to maintain backwards comparability given that we have already used this pattern quite a bit.

I'm not sure there are any cases in our current projects where we have used azure service bus QueueClients using the current pattern and actually wanted the QueueClient to be disposed of? If there are no such cases - meaning in all cases it is preferred that the QueueClient persists - then the changes made in [this commit](https://github.com/Peter554/Enable.Extensions.Queuing/commit/ed5dda95cff394eeb446e13ae6d161099f179835) may possibly be a good fix.

To be clear I'm proposing this pattern should be encouraged:

```cs
// Get a new QueueClient or request from cache if it already exists
var queue = queueClientFactory.GetQueueReference("test-queue")
// Do stuff
await queue.EnqueueAsync(message);
```

and the current pattern (below) should be deprecated. It would still be supported by making a 'dummy' Dispose method as in the commit referenced above. Note that the commit referenced presently only affects the Service Bus part of the library - while remaining compatible with the core abstraction - but similar changes could potentially be applied e.g. to the rabbitMQ part or if appropriate to the core.

```cs
// Not to be encouraged, but still technically works
using (var queue = queueClientFactory.GetQueueReference("test-queue"))
{
    await queue.EnqueueAsync(message);
}
```

The  [similar implementation](https://github.com/EnableSoftware/Enable.Extensions.Queuing/blob/master/src/Enable.Extensions.Queuing.InMemory/Internal/InMemoryQueueClient.cs) referenced may not be directly applicable as in the [Dispose method](https://github.com/EnableSoftware/Enable.Extensions.Queuing/blob/5bbc52ec412c46cac3c9e501a037139fe449fb4e/src/Enable.Extensions.Queuing.InMemory/Internal/InMemoryQueueClient.cs#L124) there we try to close a queue if the are no more references to it. This is perhaps not the desired behaviour as just because there in no-one using the queue in that particular instant it is not necessarily a good idea to close the connection as someone may want to use the queue again any arbitrarily short time after closing.

Perhaps these suggestions are bad or don't make sense - in which case no worries, please let me know.

**Comment by Peter554** [comment]
I have made changes in [this feature branch](https://github.com/Peter554/Enable.Extensions.Queuing/tree/fix/serviceBusSocketExhaustion) of my fork. These changes try to follow the linked [similar implementation](https://github.com/EnableSoftware/Enable.Extensions.Queuing/blob/master/src/Enable.Extensions.Queuing.InMemory/Internal/InMemoryQueueClient.cs) where possible, and also the solution proposed in the related issue (#21) regarding RabbitMQ.

These changes address the first two points of [this comment](https://github.com/EnableSoftware/Enable.Extensions.Queuing/issues/20#issuecomment-464737009) however I was not able to get all the current tests working reliably. I would also note that I couldn't get the tests to run reliably on master, but I may be doing something wrong.

Suggested further action:

- Get all the tests working.
- Integrate a test version of this package with an existing solution and test.
- Publish a new release version of this library to NuGet.

**Comment by Peter554** [comment]
Further inspection on master shows there have been ongoing CI build failures with similar failing tests to what I am experiencing currently (both on master and in my feature branch). For instance [this recent CI build](https://ci.appveyor.com/project/EnableSoftware/enable-extensions-queuing/builds/22457861) or one of [these](https://ci.appveyor.com/project/EnableSoftware/enable-extensions-queuing/build/0.1.38) [older](https://ci.appveyor.com/project/EnableSoftware/enable-extensions-queuing/build/0.1.39) builds. The tests fail somewhat randomly and there are often messages left on the queue when I don't expect there to be any.

I spent some time trying to fix this but didn't have much luck. I believe this to be an issue with how the tests are torn down/cleaned up and therefore there is likely nothing wrong with the **src** code. Still it would nice if we could fix these test at some point.

I have tested [my changes](https://github.com/Peter554/Enable.Extensions.Queuing/tree/fix/serviceBusSocketExhaustion) regarding the original issue locally by manual debugging and I am fairly confident they address the issue. Concurrent queue clients using the same connection string and queue-name do not make separate connections but instead reuse the cached connection. I also tried out a test version of the package with an existing solution and no breaking changes were observed.
